SimulationAgent/SimulationRunner.cs

```diff
@@ -322,84 +333,137 @@ private void UpdateDevicesStateThread()
                 }
 
                 var durationMsecs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - before;
-                this.log.Info("Device state loop completed", () => new { durationMsecs });
-                this.SlowDownIfTooFast(durationMsecs, StateLoopSettings.MIN_LOOP_DURATION);
+                this.log.Debug("Device state loop completed", () => new { durationMsecs });
+                this.SlowDownIfTooFast(durationMsecs, this.concurrencyConfig.MinDeviceStateLoopDuration);
             }
         }
 
         private void ConnectDevicesThread()
         {
+            // Once N devices are attempting to connect, wait until they are done
+            var pendingTasksLimit = this.concurrencyConfig.MaxPendingConnections;
+            var tasks = new List<Task>();
+
             while (this.running)
             {
                 this.connectionLoopSettings.NewLoop();
                 var before = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
                 foreach (var device in this.deviceConnectionActors)
                 {
-                    device.Value.Run();
+                    tasks.Add(device.Value.RunAsync());
```

Comment:
This is repeated in multiple methods.. the code can be refactored to have common helper methods.

Reply:
the code is repeated in 3 places, with one slightly different. This code is going to change considerably soon, and in my opinion the additional abstraction to deduplicate the code would make the code less readable and harder to refactor, so for the time being I will leave it as is.